openedx_tagging/core/tagging/models/base.py

```diff
@@ -260,87 +295,196 @@ def copy(self, taxonomy: Taxonomy) -> Taxonomy:
+        # Filter by search term:
+        if search_term:
+            qs = qs.filter(value__icontains=search_term)
+        qs = qs.annotate(_id=F("id"))  # ID has an underscore to encourage use of 'value' rather than this internal ID
+        qs = qs.values("value", "child_count", "depth", "parent_value", "external_id", "_id").order_by("value")
+        if include_counts:
+            # We need to include the count of how many times this tag is used to tag objects.
+            # You'd think we could just use:
+            #     qs = qs.annotate(usage_count=models.Count("objecttag__pk"))
+            # but that adds another join which starts creating a cross product and the children and usage_count become
+            # intertwined and multiplied with each other. So we use a subquery.
+            obj_tags = ObjectTag.objects.filter(tag_id=models.OuterRef("pk")).order_by().annotate(
+                # We need to use Func() to get Count() without GROUP BY - see https://stackoverflow.com/a/69031027
+                count=models.Func(F('id'), function='Count')
+            )
+            qs = qs.annotate(usage_count=models.Subquery(obj_tags.values('count')))
+        return qs
+
+    def _get_filtered_tags_deep(
+        self,
+        parent_tag_value: str | None,
+        search_term: str | None,
+        include_counts: bool,
+        excluded_values: list[str] | None,
+    ) -> TagDataQuerySet:
+        """
+        Implementation of get_filtered_tags() for closed taxonomies, where
+        we're including tags from multiple levels of the hierarchy.
+        """
+        # All tags (possibly below a certain tag?) in the closed taxonomy, up to depth TAXONOMY_MAX_DEPTH
+        if parent_tag_value:
+            main_parent_id = self.tag_for_value(parent_tag_value).pk
+        else:
+            main_parent_id = None
 
-        if not search_in_all:
-            # If not search in all taxonomy, then apply parent filter.
-            tag_set = tag_set.filter(parent=parent_tag_id)
+        assert TAXONOMY_MAX_DEPTH == 3  # If we change TAXONOMY_MAX_DEPTH we need to change this query code:
+        qs: models.QuerySet = self.tag_set.filter(
+            Q(parent_id=main_parent_id) |
+            Q(parent__parent_id=main_parent_id) |
+            Q(parent__parent__parent_id=main_parent_id)
+        )
 
         if search_term:
-            # Apply search filter
-            tag_set = tag_set.filter(value__icontains=search_term)
-
-        return tag_set.order_by("value", "id")
+            # We need to do an additional query to find all the tags that match the search term, then limit the
+            # search to those tags and their ancestors.
+            matching_tags = qs.filter(value__icontains=search_term).values(
+                'id', 'parent_id', 'parent__parent_id', 'parent__parent__parent_id'
+            )
+            if excluded_values:
+                matching_tags = matching_tags.exclude(value__in=excluded_values)
```

Comment:
Good point, but I'm going to keep it simple and just raise NotImplementedError in that case: 68f6f1995fcd5d0c5f334aee0f84b1d3d1192b16

I'm actually not sure if we even need `excluded_values` - it may be simpler to do the filtering on the frontend. But I've left it in for now.